### PR TITLE
Add experimental `cc_static_library` rule

### LINKED ISSUE #1920: Rule to build a fully static library
Among the multiple "release" outputs of my project, I'd like to include a single static library, which packages the objects of all dependencies into it. However, from what I can tell, bazel cannot build such targets.

The following target produces a single `.so` containing all symbols from `//product/foo` and `//product/bar`.:

```
cc_binary(
  name = "libproduct.so",
  linkshared = 1,
  deps = [
     "//product/foo",
     "//product/bar",
  ],
)
```

I believe there is no way to generate a similar static library (`libproduct.a`).
Is that something that will be of interest to add?

(#492 seems somewhat related, but not exactly)

### PR DESCRIPTION
**Pull request by fmeum**
RELNOTES: The new `cc_static_library` rule produces a static library that bundles given targets and all their transitive dependencies. It has to be enabled via `--experimental_cc_static_library`.

Implements https://docs.google.com/document/d/1jN0LUmp6_-rV9_f-Chx-Cs6t_5iOm3fziOklzBGjGIg/edit

Fixes #1920

### CODE REVIEW

File: tools/cpp/nm_validate_static_library.sh.tpl
Comment (aminya) [inline tools/cpp/nm_validate_static_library.sh.tpl:28]: I tried my updated branch which is based on this, and the validation failed with a lot of duplicate symbols. Not sure why there are so many duplicate symbols.

Comment (aminya) [reply tools/cpp/nm_validate_static_library.sh.tpl:28]: Interestingly, I tried stripping the static library before this script did validation, and it seems to fix the issue partially. The duplicate symbols are not outputted. 

But I get things like:
```
nm: test.pic.o: no symbols
```

**Comment by aminya** [comment]
@fmeum Any update on my comments?

**Comment by Algomorph** [comment]
@aminya and @fmeum any updates on the status / timeline for this? We (as in, my company) are considering moving off Bazel partially or completely and lack of transitive static libraries is one of the major factors in making our decisions. We need to understand our timeline and resource allocation should be and that hinges on this MR + issue thread.

**Comment by fmeum** [comment]
Sorry for the lack of response, other efforts kept needing priority over the one. I'm planning to get back to it in April.

@Algomorph If you really just want to build transitive static libraries, then you could really use any of the Starlark rules proposed on the original issue without being blocked on me. Do you have any special requirements that only my design would address?

**Comment by Algomorph** [comment]
@fmeum thanks for getting back to me so promplty and sharing the plan. April will probably work for us. I got the general sense from reading through the thread that people are running into issues with the proposed rules, but granted, I haven't personally tried any of them yet. We will review further at our next planning session and reevaluate, perhaps even assist -- if we can allocate the resources for it.

**Comment by aminya** [comment]
> We (as in, my company) are considering moving off Bazel partially or completely and lack of transitive static libraries

Unfortunately, one of my dependencies uses Bazel, and I am stuck in the walled garden of Bazel without a way out. 

> Do you have any special requirements that only my design would address?

I tried all the proposed rules, and only this PR seems to partially work for me except for the errors I mentioned here
https://github.com/bazelbuild/bazel/pull/16954#discussion_r1383347818

@fmeum 
Could you at least update the branch via https://github.com/fmeum/bazel/pull/1?

**Comment by fmeum** [comment]
> Could you at least update the branch via https://github.com/fmeum/bazel/pull/1?

@aminya Sorry, I didn't want to merge it directly due to its size and then forgot about it. I just rebased the current PR, please let me know if anything broke due to that.

File: tools/cpp/nm_validate_static_library.sh.tpl
Comment (fmeum) [reply tools/cpp/nm_validate_static_library.sh.tpl:28]: Do you happen to have an open-source project that this reproduces on?

Comment (aminya) [reply tools/cpp/nm_validate_static_library.sh.tpl:28]: Yes, the Drake library:
https://github.com/RobotLocomotion/drake/issues/16882

Comment (fmeum) [reply tools/cpp/nm_validate_static_library.sh.tpl]: Thanks, I now know why and will update the script to fix it.

Comment (fmeum) [reply tools/cpp/nm_validate_static_library.sh.tpl]: @aminya Could you try again with the current state of the PR? I fixed a bunch of issues and verified it on Protobuf.

**Comment by joelberkeley** [comment]
what versions of bazel will this be available on?

**Comment by Algomorph** [comment]
What is the current status of this MR / any blockers holding up the review from @comius? Is there anything I can do to help speed up the process?

File: src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl
Comment (comius) [inline src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl:30]: Is this the only reason, cc_static_library can go directly into rules_cc?

File: src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl
Comment (comius) [inline src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl:159]: Keep the condition consistent with the one mentioned in a comment above. I.e. `objects != None`

File: src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl
Comment (comius) [inline src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl:39]: I believe `topological` order is important here.

File: src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl
Comment (comius) [inline src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl:46]: Should be `if lib.pic_objects != None`. There are non-empty archives in C++ linking code, which have objects set to `None`. And empty ones that have objects set to `[]`.

**Review by comius — CHANGES_REQUESTED** [review]

File: src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl
Comment (fmeum) [reply src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl:30]: I can definitely send it there, but is it ready for that yet? The Unix toolchain is quite out of sync with Bazel (https://github.com/bazelbuild/rules_cc/tree/main/cc/private/toolchain) and, as far as I can tell, there is no integration test setup yet. 

I'm also not sure how to replace `cc_internal.get_artifact_name_for_category` - is `cc_internal` accessible from allowlisted repos such as `rules_cc`?

Comment (comius) [reply src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl:30]: You're right about the integration test setup, and you probably know more about how much cc toolchains are out of sync. I'd like to improve the situation.

`cc_internal.get_artifact_name_for_category` isn't accessible in any way at the moment. I'd provide something temporary for you, until we figure out the complete story of toolchain feature configuration. (Or maybe you can use some reasonable defaults?)

We had some other rules that were new and we created them directly in the depot. Examples are: py_proto_library, java_single_jar, proto_toolchain. The reason for this was, since we'd like to move the rules out of Bazel binary, anything new should be created in the repositories.

**Comment by meisterT** [comment]
cc @pzembrod

**Review by comius — CHANGES_REQUESTED** [review]

**Review by comius — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io fork 7.3.0

**Comment by iancha1992** [comment]
@fmeum Please add docstring in `src/main/starlark/tests/builtins_bzl/cc/cc_static_library/test_cc_static_library/mock_toolchain.bzl` for `mock_cc_toolchain `

**Comment by fmeum** [comment]
> @fmeum Please add docstring in `src/main/starlark/tests/builtins_bzl/cc/cc_static_library/test_cc_static_library/mock_toolchain.bzl` for `mock_cc_toolchain `

@iancha1992 Done. Despite the `src/main` prefix, this file is actually test code. Unless test code also requires doc strings, it could make sense to exclude `src/main/starlark/tests` from the linter.

**Comment by iancha1992** [comment]
> > @fmeum Please add docstring in `src/main/starlark/tests/builtins_bzl/cc/cc_static_library/test_cc_static_library/mock_toolchain.bzl` for `mock_cc_toolchain `
> 
> @iancha1992 Done. Despite the `src/main` prefix, this file is actually test code. Unless test code also requires doc strings, it could make sense to exclude `src/main/starlark/tests` from the linter.

@comius ?

**Comment by fmeum** [comment]
> @comius WDYT? The internal CL is still blocked because of it. 

I added a doc string, do I need to do anything else to unblock it?

**Comment by iancha1992** [comment]
> > @comius WDYT? The internal CL is still blocked because of it.
> 
> I added a doc string, do I need to do anything else to unblock it?

@fmeum Oh I think this time, it's 
`The function "mock_cc_toolchain" has no docstring.`

**Comment by fmeum** [comment]
> > > @comius WDYT? The internal CL is still blocked because of it.
> > 
> > 
> > I added a doc string, do I need to do anything else to unblock it?
> 
> @fmeum `The function "mock_cc_toolchain" has no docstring.` Looks like from line 101, the `def mock_cc_toolchain(name, provide_validate_static_library = True)`

@iancha1992 Should be fixed now, could you run the linter again?

**Comment by fmeum** [comment]
@bazel-io fork 7.4.0

**Comment by iancha1992** [comment]
> > > > @comius WDYT? The internal CL is still blocked because of it.
> > > 
> > > 
> > > I added a doc string, do I need to do anything else to unblock it?
> > 
> > 
> > @fmeum `The function "mock_cc_toolchain" has no docstring.` Looks like from line 101, the `def mock_cc_toolchain(name, provide_validate_static_library = True)`
> 
> @iancha1992 Should be fixed now, could you run the linter again?

I think also adding a docstring as the first statement of the file is needed. This will do. @fmeum

**Comment by fmeum** [comment]
@iancha1992 Added!

**Comment by wxfred** [comment]
Any plan to add  **cc_static_library** rule to bazel 6.x, please?

**Pull request closed without merging**
